tools/queries.cpp

```diff
@@ -44,85 +45,153 @@
+        for (auto const& query_times: times_per_query) {
+            auto sorted_query_times = query_times;
+            std::sort(sorted_query_times.begin(), sorted_query_times.end());
+            std::size_t sample_count = sorted_query_times.size();
+            double median = 0;
+            if (sample_count % 2 == 1) {
+                median = sorted_query_times[sample_count / 2];
+            } else {
+                median =
+                    (sorted_query_times[sample_count / 2] + sorted_query_times[sample_count / 2 - 1])
+                    / 2;
+            }
+            aggregated_query_times.push_back(median);
+        }
+    } else if (aggregation_type == AggregationType::Max) {
+        for (auto const& query_times: times_per_query) {
+            aggregated_query_times.push_back(*std::max_element(query_times.begin(), query_times.end()));
+        }
     }
+    std::sort(aggregated_query_times.begin(), aggregated_query_times.end());
+    return aggregated_query_times;
 }
 
-template <typename Functor>
-void op_perftest(
-    Functor query_func,
+template <typename Fn>
+void extract_times(
+    Fn query_func,
     std::vector<Query> const& queries,
     std::vector<Score> const& thresholds,
     std::string const& index_type,
     std::string const& query_type,
     size_t runs,
     std::uint64_t k,
-    bool safe
+    bool safe,
+    std::ostream* os = nullptr
 ) {
-    std::vector<double> query_times;
-    std::size_t num_reruns = 0;
-    spdlog::info("Safe: {}", safe);
+    std::vector<std::vector<std::size_t>> times_per_query(
+        queries.size(), std::vector<std::size_t>(runs)
+    );
+    std::size_t corrective_rerun_count = 0;
 
+    // Note: each query is measured once per run, so the set of queries is
+    // measured independently in each run.
     for (size_t run = 0; run <= runs; ++run) {
-        size_t idx = 0;
+        size_t query_idx = 0;
         for (auto const& query: queries) {
             auto usecs = run_with_timer<std::chrono::microseconds>([&]() {
-                uint64_t result = query_func(query, thresholds[idx]);
+                uint64_t result = query_func(query, thresholds[query_idx]);
                 if (safe && result < k) {
-                    num_reruns += 1;
+                    corrective_rerun_count += 1;
                     result = query_func(query, 0);
                 }
                 do_not_optimize_away(result);
             });
             if (run != 0) {  // first run is not timed
-                query_times.push_back(usecs.count());
+                times_per_query[query_idx][run - 1] = usecs.count();
             }
-            idx += 1;
+            query_idx += 1;
         }
     }
 
-    if (false) {
-        for (auto t: query_times) {
-            std::cout << (t / 1000) << std::endl;
-        }
-    } else {
-        std::sort(query_times.begin(), query_times.end());
-        double avg =
+    // Print JSON summary
```

Comment:
We should avoid formatting JSON by hand, we have a library for that in our deps already (`#include <nlohmann/json.hpp>`), which allows you to define JSON similar to a map, and then print it.